### ISSUE TITLE
Unexpected "permission denied" error

### ISSUE THREAD

**Issue by TristanCP** [docker/docker#30285]
**Detailed description**

I'm running an official Tomcat image (https://github.com/docker-library/tomcat/tree/master/8.5/jre8) with customized permissions applied through a Dockerfile. I've created a tomcat:tomcat user and group and made him owner of /usr/local/tomcat and all subdirectories.

Here is how I apply tomcat permissions in my Dockerfile according to OWASP recommandations :

```
FROM tomcat:8.5.11-jre8

RUN useradd --system -M --shell /bin/false --user-group tomcat

RUN chown -R tomcat:tomcat $CATALINA_HOME \
    && chmod 400 $CATALINA_HOME/conf/*.* \
    && chmod 300 $CATALINA_HOME/logs \
    && rm -Rf $CATALINA_HOME/webapps/* \
    && rm -Rf $CATALINA_HOME/server/webapps/* \
    && rm -Rf $CATALINA_HOME/conf/Catalina/localhost/*
(...)
USER tomcat
CMD ["catalina.sh", "run"]
```
Here is my full Dockerfile in case it matters : [Dockerfile](https://drive.google.com/open?id=0B5q5SPyg8HONZXNpMldBSmJhQkE)

So "chown" and "chmod" are in the same "RUN" and in expected order.

Now I connect as "tomcat" to my container :

`sudo docker exec -it 462080a55bca /bin/bash`

In /usr/local/tomcat, when I do "touch test" or "mkdir testdir", it works, but if I do the same thing in "work" subdirectory or any subdirectory, it fails (as you can see "work" is owned by tomcat and in 750) :

```
tomcat@462080a55bca:/usr/local/tomcat$ ll | grep work
drwxr-x---  2 tomcat tomcat  4096 Jan 10 21:03 work
```

```
tomcat@462080a55bca:/usr/local/tomcat$ touch work/test
touch: cannot touch ‘work/test’: **Permission denied**
```

```
tomcat@462080a55bca:/usr/local/tomcat$ mkdir work/testdir
mkdir: cannot create directory ‘work/testdir’: **Permission denied**
```
To be clear : work is NOT a mounted directory from host, just a regular directory inside the offical image with a new owner (no chmod).

**Output of `docker version`:**

```Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:38:45 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:38:45 2017
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 56
Server Version: 1.12.6
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 76
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null overlay bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 11.71 GiB
Name: koala
ID: I6XQ:72DN:WNA4:VYSQ:I2ZP:PP7D:TGCA:RHRS:XHFX:VGZT:XV3J:DUYC
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://10.91.31.45:3128
Https Proxy: http://10.91.31.45:3128
No Proxy: 127.0.0.1,localhost,.edf.fr
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running on a physical Ubuntu 16.04.1 LTS workstation.

**Comment by TristanCP** [comment]
This bug can be reproduced with Docker v1.13.0
It can NOT be reproduced on Archlinux (kernel 4.8.13)

**Comment by TristanCP** [comment]
There is a workaround : declaring ${CATALINA_HOME}/work and ${CATALINA_HOME}/work as anonymous volumes. But I don't need to persist anything between containers drop and run so it shouldn't be necessary.

**Comment by TristanCP** [comment]
This looks related to : #783 but I'm not deleting anything, just chowning an existing dir, and I can see the dirperm1 warning in dmesg info, so I guess dirperm1 option is activated for aufs but not effective.

**Comment by TristanCP** [comment]
Reproduced with overlay2
NOT reproduced with devicemapper (but well devicemapper is not for production, is it ?).

**Comment by TristanCP** [comment]
Ok, I get it.
- Workaround : switch to devicemapper storage driver (not overlay2)

- Fix : upgrade your linux kernel to >= 4.7 (tested with 4.8.0-34, fixes docker use through aufs and overlay2)

Using a kernel 4.8 means you are no more in LTS release with Ubuntu or Debian, so not appliable for production. CoreOS ships with a 4.7.3 kernel though.

So, can this be closed as "not a docker issue" ? I dont think so, not until the Docker documentation is up to date with a big warning.

**Comment by hakamairi** [comment]
That's a duplicate: https://github.com/docker/docker/issues/783
Moreover: https://github.com/docker-library/tomcat/issues/35

**Comment by justincormack** [comment]
@TristanCP many of these fixes are being backported to the stable 4.4 kernels, so this one may be at some point, for `overlay`. Filing a bug with Ubuntu might help. The new Debian stable with 4.9 kernel will be out in a few months, and Docker for AWS and Azure also have 4.9 kernels.

**Comment by TristanCP** [comment]
@justincormack It's good to know about the incoming backports, but do you know how did this start ? Was it a regression in some 4.x kernel ? or was it always there ? Because it's a big barrier for the use of Docker in production, not to be able to chown directories on official images when they are build to run as root (like the Tomcat ones).

**Comment by justincormack** [comment]
I don't believe it is a regression...

**Comment by TristanCP** [comment]
The backport of kernel 4.8 is now released with v16.04.2 LTS. I can't downgrade my 16.10, but I guess it solves this problem for Ubuntu LTS version.

**Comment by TristanCP** [comment]
After a test, I can confirm the bug is fixed (at least for aufs the default storage driver) on Ubuntu 16.04.2 LTS.

If you have upgraded to this version from a previous one, "sudo apt upgrade" won't be enough, you also need to do "sudo apt install --install-recommends xserver-xorg-hwe-16.04" to upgrade the kernel from 4.4 to 4.8
